### PR TITLE
WIP FileSource rewritten as graph stage using Java NIO

### PR DESCRIPTION
**Pull request by johanandren**
I thought it was about time for it.

### CODE REVIEW

**Comment by ktoso** [comment]
Oh cool, about time to rewrite that one :)

**Comment by johanandren** [comment]
I'll add some benchmarks to see how it differs, the old one had readahead and trixery, but I'm not convinced that is needed.

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala [hunk 1]
```diff
@@ -4,46 +4,118 @@
...
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+        } catch {
+          case ex: IllegalArgumentException ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: Hmmm, I'm not sure that's all correct. 
If the file does not exist, should we not fail the entire future?

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala [hunk 2]
```diff
@@ -4,46 +4,118 @@
...
+            throw ex
+        }
+        channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: inside try as well?

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala [hunk 3]
```diff
@@ -4,46 +4,118 @@
...
+            }
+            if (readBytes == 0 || readBytes < chunkSize) {
+              // hit end, our work here is done
+              complete(out)
+              ioResultPromise.trySuccess(IOResult(position, Success(Done)))
+            }
+
+          case Failure(ex) ⇒
+            failStage(ex)
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: yes, that's correct though

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala [hunk 4]
```diff
@@ -4,46 +4,118 @@
...
+
+        }.invoke _
+
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          channel.read(buffer, position, chunkCallback, FileSource.completionHandler)
+        }
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: good

**Review by ktoso — COMMENTED** [review]
Nice re-impl, me likes. LGTM, had one question on failing the outer future though, pretty sure we want to fail the entire materialized Future if file does not exist etc. Did it not have tests (my bad if not)

**Comment by johanandren** [comment]
I would have also liked that, but there is test coverage expecting just the inner IOResult try to fail, so I didn't dare to change that behavior. (Also the reason for the note above the cases in the try-catch)

**Comment by johanandren** [comment]
I guess the reason of the weirdness is to avoid having to deal with IO failure on two levels in the consuming code - both the `Future` failing and the possible inner `IOResult` having the failure.

**Comment by johanandren** [comment]
@ktoso you think the behaviour with regards to failures covered by the existing tests should be changed, I was thinking it will break peoples code so that's why I kept it weird. (Keep it weird, Austin!)

**Comment by agolubev** [comment]
FileInputSteam has similar functionality - it throws error if data source is not available. We have the cases - available file, empty file and not available filed. In case it's not available we can do recoverWith with alternative source. Does it make sense?

**Comment by agolubev** [comment]
New exception can be introduced- `DataSourceNotAvailable` and the reason will be child exception

**Comment by agolubev** [comment]
And, yes, - I would failure IOResult. Need some input on how this will influence compatibility and when his actually needs to be merged.

**Comment by johanandren** [comment]
I think the current behaviour is weird but I user code have to depend on the weird behaviour and if we change this it will break, so that makes me very cautious of "fixing" the behaviour to something more expected.

**Comment by agolubev** [comment]
Got it. Ok. We'll keep it as is then

File: akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala [hunk 5]
```diff
@@ -4,46 +4,118 @@
...
-  override def withAttributes(attr: Attributes): Module =
-    new FileSource(f, chunkSize, attr, amendShape(attr))
+  val out = Outlet[ByteString]("FileSource.out")
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.fileSource
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: I can't see that this is completed if downstream cancels, should it?

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala]: You are right, should be in postStop or such. I think hovewer we can close this ticket as @agolubev forked a new one based on this idea

**Comment by johanandren** [comment]
Closing as it was continued in #21801

**Pull request closed without merging**